cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/WorkflowDataset.java

```diff
@@ -0,0 +1,160 @@
+  }
+
+  private byte[] createRowKey(String key, long time) {
+    byte[] result = Bytes.toBytes(key);
+    byte[] rowKey = new byte[result.length + 8];
+    int offset = Bytes.putBytes(rowKey, 0, result, 0, result.length);
+    Bytes.putBytes(rowKey, offset, Bytes.toBytes(time), 0, Bytes.SIZEOF_LONG);
+    return rowKey;
+  }
+
+  public void write(Id.Program id, RunRecordMeta runRecordMeta, List<ActionRuns> actionRunsList) {
+
+    long start = runRecordMeta.getStartTs();
+    String key = id.getApplication().getNamespaceId() + DELIMITER
+      + id.getApplicationId() + DELIMITER + id.getId() + DELIMITER;
+    byte[] rowKey = createRowKey(key, start);
+
+    long timeTaken = runRecordMeta.getStopTs() - start;
+
+    String value = GSON.toJson(actionRunsList, ACTION_RUNS_TYPE);
+
+    table.put(rowKey, RUNID, Bytes.toBytes(runRecordMeta.getPid()));
+    table.put(rowKey, TIME_TAKEN, Bytes.toBytes(timeTaken));
+    table.put(rowKey, NODES, Bytes.toBytes(value));
+  }
+
+  public List<WorkflowRunRecord> scan(Id.Program id, long timeRangeStart, long timeRangeEnd) {
+    String namespaceId = id.getApplication().getNamespaceId();
+    String applicationId = id.getApplicationId();
+    String workflowId = id.getId();
+
+    String key = namespaceId + DELIMITER + applicationId + DELIMITER + workflowId + DELIMITER;
+    byte[] startRowKey = createRowKey(key, timeRangeStart);
+    byte[] endRowKey = createRowKey(key, timeRangeEnd);
+    Scan scan = new Scan(startRowKey, endRowKey);
+
+    Scanner scanner = table.scan(scan);
+    Row indexRow;
+    List<WorkflowRunRecord> workflowRunRecordList = new ArrayList<>();
+    while ((indexRow = scanner.next()) != null) {
+      Map<byte[], byte[]> columns = indexRow.getColumns();
+      String workflowRunId = Bytes.toString(columns.get(RUNID));
+      long timeTaken = Bytes.toLong(columns.get(TIME_TAKEN));
+
+      List<ActionRuns> actionRunsList = GSON.fromJson(Bytes.toString(columns.get(NODES)), ACTION_RUNS_TYPE);
+      WorkflowRunRecord workflowRunRecord = new WorkflowRunRecord(workflowRunId, timeTaken, actionRunsList);
+      workflowRunRecordList.add(workflowRunRecord);
+    }
+    return workflowRunRecordList;
+  }
+
+  /**
+   * Internal class to keep track of Workflow Run Records
+   */
+  public static class WorkflowRunRecord {
+    private String workflowRunId;
+    private long timeTaken;
+    private List<ActionRuns> actionRuns;
+
+    public WorkflowRunRecord(String workflowRunId, long timeTaken, List<ActionRuns> actionRuns) {
+      this.actionRuns = actionRuns;
+      this.timeTaken = timeTaken;
+      this.workflowRunId = workflowRunId;
+    }
+
+    public long getTimeTaken() {
+      return timeTaken;
+    }
+
+    public List<ActionRuns> getActionRuns() {
+      return actionRuns;
+    }
+
+    public String getWorkflowRunId() {
+      return workflowRunId;
+    }
+  }
+
+  /**
+   * Internal Class for keeping track of actions in a workflow
+   */
+  public static class ActionRuns {
```

Comment:
If that's the only place its used outside of this class, then it should be package private